pkg/integrations/v2/app_o11y_exporter/exporters/loki.go

```diff
@@ -0,0 +1,113 @@
+package exporters
+
+import (
+	"time"
+
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/go-logfmt/logfmt"
+	"github.com/grafana/agent/pkg/integrations/v2/app_o11y_exporter/models"
+	"github.com/grafana/agent/pkg/integrations/v2/app_o11y_exporter/utils"
+	loki "github.com/grafana/agent/pkg/logs"
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/logproto"
+	prommodel "github.com/prometheus/common/model"
+)
+
+// LokiExporterConfig holds the configuration of the loki exporter
+type LokiExporterConfig struct {
+	SendEntryTimeout int
+	LokiInstance     *loki.Instance
+	ExtraLabels      map[string]string
+}
+
+// LokiExporter is the struct of the loki exporter
+type LokiExporter struct {
+	li          *loki.Instance
+	seTimeout   time.Duration
+	logger      kitlog.Logger
+	extraLabels prommodel.LabelSet
+}
+
+// Init implements the AppReceiverExporter interface
+func (le *LokiExporter) Init() error {
+	return nil
+}
+
+// NewLokiExporter creates a new Loki loki exporter with the given
+// configuration
+func NewLokiExporter(logger kitlog.Logger, conf LokiExporterConfig) AppReceiverExporter {
+	el := make(prommodel.LabelSet)
+	for k, v := range conf.ExtraLabels {
+		el[prommodel.LabelName(k)] = prommodel.LabelValue(v)
+	}
+
+	return &LokiExporter{
+		logger:      logger,
+		li:          conf.LokiInstance,
+		seTimeout:   time.Duration(conf.SendEntryTimeout),
+		extraLabels: el,
+	}
+}
+
+// Export implements the AppDataExporter interface
+func (le *LokiExporter) Export(payload models.Payload) error {
+	meta := payload.Meta.KeyVal()
+
+	// log events
+	for _, logItem := range payload.Logs {
+		kv := logItem.KeyVal()
+		utils.MergeKeyVal(kv, meta)
+		le.sendKeyValsTolokiPipeline(kv)
+	}
+	// exceptions
+	for _, exception := range payload.Exceptions {
+		kv := exception.KeyVal()
+		utils.MergeKeyVal(kv, meta)
+		le.sendKeyValsTolokiPipeline(kv)
+	}
+
+	// measurements
+	for _, measurement := range payload.Measurements {
+		kv := measurement.KeyVal()
+		utils.MergeKeyVal(kv, meta)
+		le.sendKeyValsTolokiPipeline(kv)
+	}
+
+	return nil
+}
+
+func (le *LokiExporter) sendKeyValsTolokiPipeline(kv *utils.KeyVal) {
+	line, err := logfmt.MarshalKeyvals(utils.KeyValToInterfaceSlice(kv)...)
+	if err != nil {
+		level.Error(le.logger).Log("msg", "failed to logfmt a frontend log event", "err", err)
+		return
+	}
+	sent := le.li.SendEntry(api.Entry{
+		Labels: le.labelSet(kv),
+		Entry: logproto.Entry{
+			Timestamp: time.Now(),
+			Line:      string(line),
+		},
+	}, time.Duration(le.seTimeout))
+	if !sent {
+		level.Warn(le.logger).Log("msg", "failed to log frontend log event to loki pipeline")
+	}
+}
+
+func (le *LokiExporter) labelSet(kv *utils.KeyVal) prommodel.LabelSet {
```

Comment:
this is not correct: it shouldn't set all of the key value pairs as labels for the loki message. It would result in high cardinality and unnecessarily impact loki performance. User needs to explicitly configure what should be set as labels.

So, it should only set keys that are configured in `config.logsLabels`. If value is provided for a label, this value should be set directly. If value is empty string, it should be taken from entry in `kv` matching the label name. See the previous impl here: https://github.com/grafana/agent/pull/1031/files#diff-0805f060c9a440ed60d107f64f4adb402891b547a6300c3f58d6e80356c79f81